src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.OSX.cs

```diff
@@ -32,6 +37,67 @@ private unsafe int CompareStringNative(ReadOnlySpan<char> string1, ReadOnlySpan<
             return result;
         }
 
+        private unsafe int IndexOfCoreNative(ReadOnlySpan<char> source, ReadOnlySpan<char> target, CompareOptions options, int* matchLengthPtr, bool fromBeginning)
+        {
+            Debug.Assert(!GlobalizationMode.Invariant);
+            Debug.Assert(!target.IsEmpty);
+            Debug.Assert(!GlobalizationMode.UseNls);
+            Debug.Assert(target.Length != 0);
+
+            if (_isAsciiEqualityOrdinal && CanUseAsciiOrdinalForOptions(options))
+            {
+                if ((options & CompareOptions.IgnoreCase) != 0)
+                    return IndexOfOrdinalIgnoreCaseHelper(source, target, options, matchLengthPtr, fromBeginning);
+                else
+                    return IndexOfOrdinalHelper(source, target, options, matchLengthPtr, fromBeginning);
+            }
+            else
+            {
+                // GetReference may return nullptr if the input span is defaulted. The native layer handles
+                // this appropriately; no workaround is needed on the managed side.
+
+                fixed (char* pSource = &MemoryMarshal.GetReference(source))
+                fixed (char* pTarget = &MemoryMarshal.GetReference(target))
+                {
+                    NSRange result = Interop.Globalization.IndexOfNative(m_name, m_name.Length, pTarget, target.Length, pSource, source.Length, options, fromBeginning);
+                    if (matchLengthPtr == null)
+                        matchLengthPtr = &result.Length;
+                    *matchLengthPtr = result.Length;
+                    return result.Location;
+                }
+            }
+        }
```

Comment:
Since the majority of the behavior here matches ICU behavior, could we rely on the ICU code path and have a 
```
#elif TARGET_OSX || TARGET_MACCATALYST || TARGET_IOS || TARGET_TVOS
                if (GlobalizationMode.Hybrid)
                {
                    ....
#endif
```
like what was done for `IndexOfOrdinalIgnoreCaseHelper` and `IndexOfOrdinalHelper`?

And same comments about the `matchLengthPtr` and possible modification to `Interop.Globalization.IndexOfNative` as suggested above